pkg/operator/controllers/checkers/serviceprincipalchecker/checker.go

```diff
@@ -22,18 +22,14 @@ type servicePrincipalChecker interface {
 type checker struct {
 	log *logrus.Entry
 
-	credentials        func(ctx context.Context) (*clusterauthorizer.Credentials, error)
-	getTokenCredential func(azEnv *azureclient.AROEnvironment, credentials *clusterauthorizer.Credentials) (azcore.TokenCredential, error)
+	getTokenCredential func(azEnv *azureclient.AROEnvironment) (azcore.TokenCredential, error)
```

Comment:
what's the purpose of this as a struct?  So that the controller can fetch a new token credential every time it needs to run the checker, rather than once on startup?

Reply:
I'm also wondering if this is "legacy".  The newer azure-sdk-for-go no longer requires us to refresh the token iirc.

Reply:
>what's the purpose of this as a struct? So that the controller can fetch a new token credential every time it needs to run the checker, rather than once on startup?

I'm not 100% sure, but it seems that way to me.

>I'm also wondering if this is "legacy". The newer azure-sdk-for-go no longer requires us to refresh the token iirc.

I wasn't aware of this. I'll have to look into it a bit, and then maybe we can remove unneeded stuff from the struct.

Reply:
After some digging, I found that you're right - as long as we're using azidentity, our tokens will get auto-refreshed before they expire. So I'll see about refactoring this struct a bit.

For context, https://github.com/Azure/ARO-RP/pull/2667 switched us over to `azidentity`, which uses MSAL under the hood, and https://learn.microsoft.com/en-us/entra/identity-platform/msal-overview states that MSAL refreshes tokens automatically.

Reply:
Well, after looking at this code in more detail and considering how we might refactor it, I think we might as well leave it how it is.

It looks like for all of the checkers under `pkg/operator/controllers/checkers`, there's a pattern of defining a `SomethingChecker` interface and then creating a struct type that implements the interface. So it logically makes sense to have that struct type encapsulate anything related to the `Check` that the controller in question does. I think `getTokenCredential` and `newSPValidator` are included in the struct more because its logical than because of a goal of getting a new token each time the controller reconciles (which is what I was originally thinking).

Additionally, if you look at the storage accounts controller for example (specifically https://github.com/Azure/ARO-RP/blob/ac7cf2f35c24c1e1d39960cb1b382de7def8f9ba/pkg/operator/controllers/storageaccounts/storageaccount_controller.go#L95), it also "manually" creates a new token credential each time it `Reconcile`s. So if we wanted to refactor to only create a token credential on controller startup, we'd probably want to do that to all of the controllers that leverage Azure clients and not just the service principal checker.

Let me know what you think!